packages/backend/src/document.rs

```diff
@@ -123,3 +125,192 @@ pub struct RefContent {
+
+    // selects the ref id and ref name (from the most recent snapshot of that
+    // ref) for all refs that a user (the searcher) has permission to access.
+    // Optionally filter the results by the owner of the refs.
+    let results = sqlx::query!(
+        r#"
+        WITH latest_snapshots AS (
+            SELECT DISTINCT ON (snapshots.for_ref) 
+                refs.id AS ref_id, 
+                refs.created::TEXT AS created_at,
+                snapshots.content->>'name' AS name,
+                snapshots.content->>'type' AS type_name
+            FROM snapshots
+            JOIN refs ON snapshots.for_ref = refs.id
+            ORDER BY snapshots.for_ref, snapshots.id DESC
+        )
+        SELECT 
+            ls.ref_id,
+            ls.name,
+            ls.type_name,
+            ls.created_at,
+            p_searcher.level as "permission_level: PermissionLevel",
+            owner.id as owner_id,
+            owner.username as owner_username,
+            owner.display_name as owner_display_name
+        FROM latest_snapshots ls
+        JOIN permissions p_searcher ON ls.ref_id = p_searcher.object
+        LEFT JOIN users owner ON owner.id = (
+            SELECT p_owner.subject 
+            FROM permissions p_owner 
+            WHERE p_owner.object = ls.ref_id 
+            AND p_owner.level = 'own' 
+            LIMIT 1
+        )
+        WHERE (
+            (p_searcher.subject = $1
+            AND p_searcher.level IN ('read', 'write', 'maintain', 'own'))
+            OR p_searcher.subject IS NULL -- user is allowed to search public documents
+        )
+        AND (
+            owner.username = $2  -- owner_username
+            OR $2 IS NULL  -- include all owners if owner_username is NULL
+        )
+        AND (
+            ls.name ILIKE '%' || $3 || '%'  -- case-insensitive substring search
+            OR $3 IS NULL  -- include all if name filter is NULL
+        )
+        LIMIT 100; -- TODO: pagination
+        "#,
+        searcher_id,
+        search_params.owner_username_query,
+        search_params.ref_name_query
+    )
+    .fetch_all(&ctx.state.db)
+    .await?;
+
+    // Map SQL query results to Vec<RefStub>
+    let stubs = results
+        .into_iter()
+        .map(|row| RefStub {
+            ref_id: row.ref_id,
+            name: row.name.unwrap_or_else(|| "untitled".to_string()),
+            type_name: row.type_name.expect("type_name should never be null"),
+            permission_level: row.permission_level,
+            created_at: row.created_at.expect("created_at should never be null"),
+            owner: UserSummary {
+                id: row.owner_id,
+                username: row.owner_username,
+                display_name: row.owner_display_name,
+            },
+        })
+        .collect();
+
+    Ok(stubs)
+}
+
+/// Gets a list of RefStubs that a user has permissions for and match the search parameters
+pub async fn search_ref_stubs_related_to_user(
```

Comment:
This function is similar enough to the previous one that:

- I can't easily tell from the docstrings how they are different (but the difference *was* clear from the PR comments, so thanks for explaining it there!)

- I wonder whether the two functions should be consolidated into a single function, say by adding an extra field `permission_level: Option<PermissionLevel>` to `RefQueryParams` that would restrict the results to documents for which the searcher has *at least* that level (set individually, not inherited from global permissions). So, if I understand correctly, in one case you pass `Some(Read)` or `Some(Own)` (depending on how strongly the document is supposed to be "associated" with the user) and in the other case, you pass `None`.